### PR TITLE
 React Redux SpaceX App: Data Fetching, Reservations, and Profiles

### LINKED ISSUE #3: [2pt]  Display reserved rockets in My profile - Filters
Render a list of all reserved rockets (use `filter()`) on the "My profile" page.

### LINKED ISSUE #5: [1pt] Switch badges for Rockets - Conditional components
Rockets that have already been reserved should show a "Reserved" badge and "Cancel reservation" button instead of the default "Reserve rocket" (as per design)

Use the React conditional rendering syntax:
```javascript
{rocket.reserved && ( 
    // render Cancel Rocket button
)}
```

### LINKED ISSUE #8: [1pt] Implement rocket booking cancelation - Actions
- Follow the same logic as with the "Reserve rocket" - but you need to set the `reserved` key to `false`. 
- Dispatch these actions upon click on the corresponding buttons.

### LINKED ISSUE #12: [3pt] Implement rocket booking - Actions
- When a user clicks the "Reserve rocket" button, action needs to be dispatched to update the store. You need to get the ID of the reserved rocket and update the state. Remember you mustn't mutate the state. Instead, you need to return a new state object with all rockets, but the selected rocket will have an extra key `reserved` with its value set to `true`. You could use a JS `filter()` or `map()` to set the value of the new state - i.e.:
```javascript
const newState = state.map(rocket => {
    if(rocket.id !== id) 
        return rocket;
    return { ...rocket, reserved: true };
});

```

- Regardless of which method you choose, make sure you place all your logic in the reducer. In the React view file, you should only dispatch the action with the correct rocket ID as an argument.

### LINKED ISSUE #13: [4pt] Display rockets - Lists render
- Use `useSelector()` Redux Hook to select the state slices and render lists of rockets in corresponding routes. i.e.:
```javascript
// get rockets data from the store
const rockets = useSelector(state => state.rockets);
```

- You can style the whole application "by hand" or you could use [React Bootstrap](https://react-bootstrap.github.io/), a UI library that could speed up the process. This is a popular library and working with its components would be good practice.
- Render a list of rockets (as per design). For the image of a rocket use the first image in the array of `flickr_images`.

### LINKED ISSUE #17: [4pt]  Fetch rockets - Fetch data
Fetch data from the Rockets endpoint  (https://api.spacexdata.com/v3/rockets) when the application starts (as Rockets is the default view).

Once the data are fetched, dispatch an action to store the selected data in Redux store:

- id
- name
- type
- flickr_images

NOTE: Make sure you only dispatch those actions once and do not add data to store on every re-render (i.e. when changing views / using navigation).

### LINKED ISSUE #21: [1pt] Create basic structure for Rockets - Setup
- Create a route and a view component. Use `<NavLink />` for the page navigation links and style active class to indicate which section/page the user is currently on (underline active navigation link).
- Create a directory for all Redux state slice files.

### PR DESCRIPTION
**Pull request by ahmidou09**
This pull request introduces key features to the React SpaceX App, enhancing its functionality and user experience. 

## Create basic structure for Rockets - Setup

- Create a route and a view component. Use `<NavLink />` for the page navigation links and style active class to indicate which section/page the user is currently on (underline active navigation link).
- Create a directory for all Redux state slice files.

## Fetch rockets - Fetch data

1. Fetch data from the Rockets endpoint  (https://api.spacexdata.com/v3/rockets) when the application starts (as Rockets is the default view).

1. Once the data are fetched, dispatch an action to store the selected data in Redux store:

- id
- name
- type
- flickr_images

## Display rockets - Lists render 

- Use `useSelector()` Redux Hook to select the state slices and render lists of rockets in corresponding routes:
- Render a list of rockets (as per design). For the image of a rocket use the first image in the array of `flickr_images`.

## Implement rocket booking - Actions
- When a user clicks the "Reserve rocket" button, action needs to be dispatched to update the store. You need to get the ID of the reserved rocket and update the state. Remember you mustn't mutate the state. Instead, you need to return a new state object with all rockets, but the selected rocket will have an extra key `reserved` with its value set to `true`. You could use a JS `filter()` or `map()` to set the value of the new state

## Implement rocket booking cancelation - Actions

- Follow the same logic as with the "Reserve rocket" - but you need to set the `reserved` key to `false`. 
- Dispatch these actions upon click on the corresponding buttons.

## Switch badges for Rockets - Conditional components 

- Rockets that have already been reserved should show a "Reserved" badge and "Cancel reservation" button instead of the default "Reserve rocket"

## Display reserved rockets in My profile - Filters

- Render a list of all reserved rockets (use `filter()`) on the "My profile" page.

### CODE REVIEW

**Comment by victormc13** [comment]
**Under review...**

File: src/redux/rockets/rocketsSlice.js [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+    const response = await axios.get('https://api.spacexdata.com/v4/rockets');
+    return response.data;
+  } catch (error) {
+    throw Error('Failed to fetch rocket data');
+  }
+});
+
+const initialState = {
+  rockets: [],
+  status: 'idle',
```
Comment (victormc13) [inline src/redux/rockets/rocketsSlice.js:15]: You could use boolean values for status, to make conditional rendering later, for example for missions I did:

image

File: src/redux/rockets/rocketsSlice.js [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+      .addCase(fetchRockets.fulfilled, (state, action) => {
+        state.status = 'succeeded';
+        state.rockets = action.payload;
+      })
+      .addCase(fetchRockets.rejected, (state, action) => {
+        state.status = 'failed';
+        state.error = action.error.message;
+      });
+  },
+});
```
Comment (victormc13) [inline src/redux/rockets/rocketsSlice.js]: Here we could just left all the logic and rendering for components, for exmaple (state.Loading and state.error):

image

finally in our component we can use conditional rendering like this:

image

**Review by victormc13 — CHANGES_REQUESTED** [review]
# Changes Required ♻️⚒️  

Hi @ahmidou09 

Great work on making the project requirements 👏! I really appreciate the effort you put in!  **_However, there are few things that we can to polish. But you are almost there!_** 💪 🏁

so close


## Highlights 🎯

There are some excellent points that you have made perfectly. Keep up the great work! 👏
- ✅ Good job with the structure
- ✅ Great job with name variables
- ✅ Correct Linters

## Required Changes ♻️ 

_Please check the comments under the review._

## Optional suggestions ❕
Cheers, and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag @victormc13  in your question so I can receive the notification.

File: src/redux/rockets/rocketsSlice.js (hunk 2 above)
Comment (ahmidou09) [reply src/redux/rockets/rocketsSlice.js]: By using these specific string values for the status property, your Redux reducers can easily convey the current state of asynchronous operations, making it clearer how the application is handling these operations and rendering UI components accordingly.

For further information, please refer to the [official documentation.
](https://redux-toolkit.js.org/api/createAsyncThunk)

Comment (ahmidou09) [reply src/redux/rockets/rocketsSlice.js]: The string literal type 'idle' | 'pending' | 'succeeded' | 'failed', It represents a type that can have one of four specific string values: 'idle', 'pending', 'succeeded', or 'failed'.

By using these specific string values for the status property, your Redux reducers can easily convey the current state of asynchronous operations, making it clearer how the application is handling these operations and rendering UI components accordingly.

For further information, please refer to the o[fficial documentation](https://redux-toolkit.js.org/api/createAsyncThunk).

**Review by victormc13 — APPROVED** [review]
# Approved 👏  
### **Highlights 🚀 :heavy_check_mark: :heavy_check_mark:** 

- The repo looks professional.:heavy_check_mark:
- Git/GitHub flow best practices have been followed:heavy_check_mark:
- You have descriptive PR:heavy_check_mark:
- Your project meets all the requirements. :heavy_check_mark: 

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
### **Congratulations!** 🎉 
image



Cheers, and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag @victormc13  in your question so I can receive the notification.

**Pull request merged**